.teamcity/src/main/kotlin/projects/FunctionalTestProject.kt

```diff
@@ -1,24 +1,64 @@
 package projects
 
+import common.Os
+import configurations.BaseGradleBuildType
 import configurations.FunctionalTest
 import jetbrains.buildServer.configs.kotlin.v2019_2.Project
 import model.CIBuildModel
 import model.FunctionalTestBucketProvider
 import model.Stage
 import model.TestCoverage
+import model.TestType
+import model.getBucketUuid
+
+const val DEFAULT_FUNCTIONAL_TEST_BUCKET_SIZE = 50
+const val DEFAULT_LINUX_FUNCTIONAL_TEST_BUCKET_SIZE = 20
 
 class FunctionalTestProject(
-    model: CIBuildModel,
+    val model: CIBuildModel,
     functionalTestBucketProvider: FunctionalTestBucketProvider,
     val testCoverage: TestCoverage,
-    stage: Stage
+    val stage: Stage
 ) : Project({
     this.id(testCoverage.asId(model))
     this.name = testCoverage.asName()
 }) {
-    val functionalTests: List<FunctionalTest> = functionalTestBucketProvider.createFunctionalTestsFor(stage, testCoverage)
+    val functionalTests: List<BaseGradleBuildType> = functionalTestBucketProvider.createFunctionalTestsFor(stage, testCoverage).let {
+        addDummyBuckets(it)
+    }
 
     init {
         functionalTests.forEach(this::buildType)
     }
+
+    /**
+     * These build configurations are used to retain build histories. For example, we had 50 buckets before,
+     * if we remove 30 of them, the histories are not accessible anymore. As a workaround, we don't remove
+     * the 30 buckets but not dependencies of trigger build.
+     */
+    private fun addDummyBuckets(functionalTestBuckets: List<FunctionalTest>): List<BaseGradleBuildType> {
```

Comment:
Like we discussed, the empty buckets should keep at least a dependency on `compileAll`, so the branches which have not yet been updated continue working.